src/sample.rs

```diff
@@ -0,0 +1,330 @@
+/// - Whos items can be `cloned`.
+///
+/// If cloning is impossible or expensive, use `sample_ref` instead.
+pub trait Sample {
+    /// The returned sampled data. Typically the either a `Vec<T>` or a new instance of the
+    /// container's own type.
+    type Sampled;
+
+    /// Return exactly `amount` randomly sampled values.
+    ///
+    /// Any type which implements `sample` should guarantee that:
+    /// - Both the order and values of `Sampled` is random.
+    /// - The implementation uses `O(amount)` speed and memory
+    /// - The returned values are not references (if so, implement `SampleRef` instead).
+    ///
+    /// Panics if `amount > self.len()`
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use rand::{thread_rng, Sample};
+    ///
+    /// let mut rng = thread_rng();
+    /// let values = vec![5, 6, 1, 3, 4, 6, 7];
+    /// println!("{:?}", values.sample(&mut rng, 3))
+    /// ```
+    fn sample<R: Rng>(&self, rng: &mut R, amount: usize) -> Self::Sampled;
+}
+
+/// The `SampleRef` trait provides the `sample_ref` method.
+///
+/// This is intended to be implemented for containers that which can be sampled in `O(amount)` time
+/// and want a fast way to give references to a sample of their items.
+pub trait SampleRef {
+    /// The returned sampled data. Typically the either a `Vec<&T>` or a new instance of the
+    /// container's own type containing references to the underlying data.
+    type SampledRef;
+
+    /// Return exactly `amount` references to randomly sampled values.
+    ///
+    /// Any type which implements `sample_ref` should guarantee that:
+    /// - Both the order and values of `SampledRef` is random.
+    /// - The implementation uses `O(amount)` speed and memory.
+    /// - The returned values are not copies/clones (if so, implement `Sample` instead).
+    ///
+    /// Panics if `amount > self.len()`
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use rand::{thread_rng, SampleRef};
+    ///
+    /// let mut rng = thread_rng();
+    /// let values = vec![5, 6, 1, 3, 4, 6, 7];
+    /// println!("{:?}", values.as_slice().sample_ref(&mut rng, 3))
+    /// ```
+    fn sample_ref<R: Rng>(&self, rng: &mut R, amount: usize) -> Self::SampledRef;
+}
+
+/// Randomly sample *up to* `amount` elements from a finite iterator using a reservoir.
+///
+/// The order of elements in the sample is not random. In fact, if `len(iterable) <= amount` then
+/// the output will be in the exact order they were collected.
+///
+/// The reservoir method used allocates only an `Vec` of size `amount`. The size of the iterable
+/// does not affect the amount of memory used.
+///
+/// # Example
+///
+/// ```rust
+/// use rand::{thread_rng, sample_reservoir};
+///
+/// let mut rng = thread_rng();
+/// let sample = sample_reservoir(&mut rng, 1..100, 5);
+/// println!("{:?}", sample);
+/// ```
+pub fn sample_reservoir<T, I, R>(rng: &mut R, iterable: I, amount: usize) -> Vec<T>
```

Comment:
Shouldn't this be a special case of your traits now, i.e. `impl<I: Iterator> Sample for I`?

Reply:
as said above, `sample_reservoir` is far slower.

We *could* use `sample_reservoir` for iterators or "generic" types like HashMaps, but we will not be able to guarantee performance of `O(m)`.

On that note, I've investigated samplign from HashMaps in `O(m)` and I don't think it can be done. Even the iterator represenation is *required* to iterate over the full `capacity` in order to guarantee selection of all elements -- it has no way to select *only* the "filled buckets".

Reply:
Also, anything done without creating *an ordered view* of items in the HashMap will not be reproducible, even with a reproducible generator. Theoretical and maybe not important.

Reply:
Ya, I think special-casing HashMap is not really possible, at least without architecting a HashMap for just that purpose (in which case... provide your own sampling function for it since you are creating it for that purpose).

For standard hashmap's someone will have to use `sample_reservoir` and eat the `O(N)` computational cost.